### PR TITLE
respect build_by_default and install flags

### LINKED ISSUE #12751: build_by_default isnt respected when install is true during compilation
**Describe the bug**
When a build target has both `build_by_default : false` and `install : true` running `meson compile -C build` builds the target.

**To Reproduce**
```py
project('temp', 'c')

lib = library('temp', 'temp.c',
  install : true,
  build_by_default : false
)
```
```c
// temp.c
int hello(int i) {
    return i * 10;
}
```
```sh
meson setup build
meson compile -C build # ninja -C build has the same issue
```
**Expected behavior**
`temp` should only be built during installation

**system parameters**
* windows 11 pro build 22621
* meson 1.3.1
* ninja 1.11.1
* python 3.11.7

### PR DESCRIPTION
**Pull request by apache-hb**
This fixes #12751 and #12530.

Added a new `all-install` rule that contains all the targets that are to be installed, this way installing builds required targets and the default rule doesnt pick up on them

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1261,7 +1261,7 @@ def generate_coverage_legacy_rules(self, gcovr_exe: T.Optional[str], gcovr_versi
     def generate_install(self):
         self.create_install_data_files()
         elem = self.create_phony_target('install', 'CUSTOM_COMMAND', 'PHONY')
-        elem.add_dep('all')
+        elem.add_dep('all-install')
```
Comment (bruchar1) [inline mesonbuild/backend/ninjabackend.py:1264]: I guess that if a new target name is added, it should be added to `FORBIDDEN_TARGET_NAMES` in coredata module.

File: mesonbuild/backend/backends.py [hunk 2]
```diff
@@ -1400,6 +1400,15 @@ def get_build_by_default_targets(self) -> 'T.OrderedDict[str, T.Union[build.Buil
                 result[name] = b
         return result
 
+    def get_install_targets(self) -> 'T.OrderedDict[str, T.Union[build.BuildTarget, build.CustomTarget]]':
```
Comment (bruchar1) [inline mesonbuild/backend/backends.py:1403]: You don't need to quote the annotations because annotations are imported from future

**Comment by eli-schwartz** [comment]
> this way installing builds required targets and the default rule doesnt pick up on them

Very creative. However, there are many downstream users of meson that run this:
```
ninja -C builddir

meson install -C builddir --no-rebuild
```
Rationale being that they explicitly don't want to run the compiler during the install stage, especially as it may run with different permissions (root) compared to the build stage.

This PR would break that. It also goes against the grain where running `make` or `ninja` is typically expected to perform all necessary work in advance of `make install` or `ninja install`.

Users would be unable to successfully run meson install --no-rebuild using either approach, since depending on your meson version, `ninja all` may not do the right thing, or else `ninja all-install` doesn't exist and errors out.

**Comment by eli-schwartz** [comment]
It's not clear to me this is a problem and needs to be solved.

**Pull request closed without merging**